### DISCUSSION TITLE
Action<IContainer> strange behavior

### DISCUSSION THREAD

**Discussion by Guilherme-Cassel** [QuestPDF/QuestPDF#773] [Q&A] [answered]
![image](https://github.com/QuestPDF/QuestPDF/assets/150675837/39615926-8199-48ca-a57f-b8414473e45c)

I've been trying to create a Action ColumnDescriptor (materialTable) and mount its items before applying it to the page content, but I've been incountering some strange errors and even super crazy behaviors I've never imagine could be possible on debugging such as like code GOING BACKWARDS haha.

does someone know what I'm doing wrong?

as soon as I add the "item" to the materialTable, It breakes, and the output on the actual pdf is just pure LSD

**Comment by girlpunk** [comment]
At a guess, `item` is not what you expect it to be, and will always have the last set value when the actions in `materialTable` are actually run. You could try copying `i` to a local variable inside the action, and calling `TableItem()` directly with that value from inside the action?

i.e.
```csharp
for (int i = 0; i < materialProducts.Count - 1; i++)
{
    string color = "...";
    materialTable += x =>
    {
        var index = i;
        TableItem(x.Item(), index);
    };
}

**Comment by MarcinZiabek** [comment, marked as answer]
The problem is related to the scope of the `item` variable. 

You have declared this variable before the for loop, so it exists as a single reference and is updated by each loop iteration. Lambda expression captures this variable, but as declared in the outer scope, it will store only the last item value from a lambda perspective.

However, if you put your variable inside a loop, its scope will differ, and a separate reference will be created for each loop iteration. Therefore, your lambda will capture the correct expected value (and you will likely get a warning about capturing variables and why it may introduce some unexpected pressure on GC).

You have posted a screenshot instead of code, so it is challenging to show the difference 😀

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
